SEP-0012.md

```diff
@@ -0,0 +1,495 @@
+A lot of the other changes presented in this SEP are a consequence of this decision.
+
+
+## Summary of Changes from ndcube 1
+
+The fundamental change since the original version of ndcube is that the WCS type has been changed to be the High Level API as described by astropy's APE 14.
+This opens a lot of possibilities to the future uses of `NDCube` as the APE 14 API allows the coordinate system to be described by almost any functional transform.
+This could be the most common FITS WCS standard, or gWCS or custom implementations for specific facilites with different requirements.
+
+Having made this change, which was a large breaking change from the original API, we reviewed all the existing functionality and API as it was adapted to use the high level WCS interface and the result is presented below.
+
+# Detailed Description
+
+The API presented here is limited to the two core classes `NDCubeABC` and `ExtraCoordsABC`.
+This is primarily an attempt to limit the discussion in this SEP to the core API and not implementation decisions taken in the development of ndcube.
+If people are interested in the implementation of the proposed API large amounts of it are implemented on the master branch of the [ndcube repository](https://github.com/sunpy/ndcube) or open pull requests.
+
+The `NDCube` class inherits the astropy [`NDDataBase`](https://github.com/astropy/astropy/blob/master/astropy/nddata/nddata_base.py#L11) class, including the requirement that the `.wcs` property be a `BaseHighLevelWCS` instance.
+In the following code snippets of the API use the Python type hint syntax as appropriate to highlight accepted types and return types, these are not enforced at runtime.
+
+
+## NDCube
+
+Below is a class definition for the Abstract Base Class (ABC) for `NDCube`.
+Its purpose is to specify the API as completely as possible, including input and return types. 
+The behaviour of all the methods should be described by their included docstrings, a version of this (likely out of sync) has been [PRed to ndcube](https://github.com/sunpy/ndcube/pull/304).
+
+```python
+
+class NDCubeABC(astropy.nddata.NDDataBase):
+
+    @abc.abstractmethod
+    def __init__(self,
+                 data: "numpy.typing.ArrayLike",
+                 uncertainty: Any = None,
+                 mask: Any = None,
+                 wcs: Union[BaseLowLevelWCS, BaseHighLevelWCS] = None,
+                 meta: Mapping = None,
+                 unit: Union[str, u.Unit] = None,
+                 copy: bool = False,
+                 extra_coords: "ExtraCoordsABC" = None):
+        pass
+
+    @abc.abstractmethod
+    def __getitem__(self, item: Union[int, slice, Iterable[Union[slice, int]]]) -> "NDCubeABC":
+        """
+        An ndcube object must be sliceable in a similar way to a numpy array,
+        however, support for a `step` is not required and no "fancy indexing" is allowed.
```

Comment:
- Why isn't support for a step required?
- What is fancy indexing, and again why isn't it allowed?

Reply:
>     * Why isn't support for a step required?

Because it's very non-trivial to support slicing a WCS with a step in a predictable way. A step array slicing operation does not nicely map to a single operation in world space. Are you asking for a resample operation where the resulting single pixel has the world coordinates of the whole step (i.e. new_cdelt = step * old_cdelt) or are you asking to discretise the WCS so that it has gaps in it?

It is left as "not required" here as a specific subclass of NDCube may be able to make limiting assumptions based on it's usecase and know how to implement it.

> 
>     * What is fancy indexing, and again why isn't it allowed?

Fancy indexing is https://numpy.org/doc/stable/reference/arrays.indexing.html#advanced-indexing and it's not supported because again it's practically impossible to support without converting the WCS to a full lookup table and also not all array types (Dask) support it either.

Reply:
👍 I think a (condensed?) version of the above comment should go into the docstring